Add channel descriptions on long press

jlanesmith opened the pull request:
When long pressing a channel from the Home menu, a modal will appear with the channel's name and description. 

Current questions:
- I based the style of channel modals off of UserProfileStyle.js. Should I create a new style file that is very similar, or rename it to UserChannelProfileStyle.js, or leave it as it is?
- Should the channel description be left or center aligned?
- I think I should add an info button beside the channel name because people won't know to long press. Should I put it beside the red arrow, or in place of the red arrow? And, what icon should I use?

scholvat commented:
This is great!

Usually we create separate stylesheets for each component. If multiple components share a common attribute (such as color, border, etc.), there is a global stylesheet that can be used to make future changes easier.

A nice to have on PRs is some screenshots, since not everyone has the dev environment setup, and it also makes it easier to do reviews on mobile (like I am doing now).

I'll do a full review when I get home. I took a brief read through the code and it looks pretty good.

jlanesmith commented:
![image](https://user-images.githubusercontent.com/26492359/69000682-91f91b00-08a1-11ea-8fa3-cfc2d1a1fc0d.png)

scholvat commented:
@CharlotteBaker316 comments on the UI?

CharlotteBaker316 commented:
Looks cool! Thanks for working on this! In terms of UI it's great that it currently matches our other pop-ups. For gestures, what if we included this on the channel's page itself? if we added a little info button? like this:
![skybunktemp](https://user-images.githubusercontent.com/40967355/69008280-b5aa7880-0916-11ea-81e2-1b17112daeb4.JPG)

jlanesmith commented:
@CharlotteBaker316 sounds good! Should I include the "i" on the home page as well, beside the channel name, or just in the channel's page?

scholvat commented:
> @CharlotteBaker316 sounds good! Should I include the "i" on the home page as well, beside the channel name, or just in the channel's page?

Personally I think adding the icon to the home page will add a lot of clutter. But that's not a strong opinion, I think either way is fine.

CharlotteBaker316 commented:
I agree, I don’t think it needs to be on the channel page. I think just the page itself is great!

CharlotteBaker316 commented:
And maybe if at some point it could get integrated into where users make posts, that could be cool too.

picklechips commented on views/Home/Home.jsx:
```diff
@@ -108,6 +111,19 @@ export default class HomeView extends React.Component {
 
+  onLongPressChannel = (channelId, channelName) => {
+    const { channels, user } = this.state;
+    var channel;
```
using `let` instead of `var` is a more modern syntax

picklechips commented on views/Home/Home.jsx:
```diff
@@ -108,6 +111,19 @@ export default class HomeView extends React.Component {
+    var channel;
+
+    if (['all', 'subs', 'myPosts'].includes(channelId)) channel = { _id: channelId, name: channelName };
+    else channel = _.head(_.filter(channels, { _id: channelId }));
```
This can be done without lodash using es6 syntax (I don't really like that we have lodash as a dependency)
```
channel = channels.find(channel => channel._id === channelId)
```

picklechips commented on components/ChannelProfile/ChannelProfile.jsx:
```diff
@@ -0,0 +1,69 @@
+    onClose();
+  }
+
+  getCloseJSX() {
```
Personally I would just put this JSX right in the render function since there isn't any additional logic being done to generate the jsx

picklechips commented on components/ChannelProfile/ChannelProfile.jsx:
```diff
@@ -0,0 +1,69 @@
+import { Text, Icon, Card, CardItem } from 'native-base';
+import { Font, AppLoading } from "expo";
+import AutoLink from 'react-native-autolink'
+import _ from 'lodash';
```
Looks like not all of these are actually being used, so it's probably best to remove them.

eg. Font, AppLoading, lodash

picklechips commented on views/Home/Home.jsx:
```diff
@@ -108,6 +111,19 @@ export default class HomeView extends React.Component {
+    const { channels, user } = this.state;
+    var channel;
+
+    if (['all', 'subs', 'myPosts'].includes(channelId)) channel = { _id: channelId, name: channelName };
```
Not directly related to this PR but just making a general note:
It might make sense to just append these channels to `state.channels` here instead of in `ChannelList.jsx` so we don't have to do this custom logic

picklechips replied:
Actually maybe that's a smell that this logic should be in ChannelList, since Home probably shouldn't need to known about these channels. I don't see any reason why this method can't be defined in `ChannelList.jsx`.

No strong opinion though

jlanesmith replied:
I'm not really familiar with this, so I'm gonna leave this for another PR, if that's okay

jlanesmith commented:
@scholvat @picklechips I'm trying to add the info button as Charlotte suggested, but I'm encountering a problem. The navigation options are static, but the channel description modal uses class functions and variables. If I want to add the info button to the navigation bar, it appears that it needs to be static, but that means I'll have to make the channel modal static, which I don't think is a good idea. Do you have any suggestions?

scholvat commented:
Have you looked at EditProfile view? I forget exactly how it works, but I think that is a decent example of how to pass data from the navigation bar to the render function.

jlanesmith commented:
Here's what the icon looks like in the feed
![image](https://user-images.githubusercontent.com/26492359/70378803-a786c100-18f2-11ea-9af5-7fc0b12c1644.png)

jlanesmith commented:
@scholvat @picklechips I think I've addressed all your comments. However, I do have a question regarding CSS. For channels where the description is more than 1 line, it centers properly but for channels where the description is only 1 line, it is left aligned. In `ChannelProfileStyle.js`, I set `textAlign: 'center'` and I can't figure out why it doesn't always work. Do you have any ideas?

scholvat commented:
> @scholvat @picklechips I think I've addressed all your comments. However, I do have a question regarding CSS. For channels where the description is more than 1 line, it centers properly but for channels where the description is only 1 line, it is left aligned. In `ChannelProfileStyle.js`, I set `textAlign: 'center'` and I can't figure out why it doesn't always work. Do you have any ideas?

Weird, it looks centered on iOS. I have no idea, sounds like a react-native bug.

picklechips commented:
> @scholvat @picklechips I think I've addressed all your comments. However, I do have a question regarding CSS. For channels where the description is more than 1 line, it centers properly but for channels where the description is only 1 line, it is left aligned. In `ChannelProfileStyle.js`, I set `textAlign: 'center'` and I can't figure out why it doesn't always work. Do you have any ideas?

What happens if you set `width: '100%'` or `flex: 1` in `channelInfoText`? If I were to hazard a guess the Text element's width is only growing with the length of the text. So if the text doesn't span the full width of the modal it appears left aligned.
Here's a diagram explaining what I'd expect is happening:
```
1 line text
--------------------
|                  |
| |text div|       |
|                  |
--------------------

multi-line text
--------------------
| |text div now  | |
| |spans full    | |
| |width so looks| |
| |centered      | |
--------------------
```

jlanesmith commented:
@picklechips Adding `flex: 1` doesn't make a difference. Adding `width: '100%'` doesn't affect the single line text, but instead makes the multiline descriptions extend outside of the box, instead of wrapping around like they are now.

picklechips commented:
> Adding width: '100%' doesn't affect the single line text, but instead makes the multiline descriptions extend outside of the box

Lol that definitely shouldn't happen - something must be wacky with the underlying modal CSS. Anyhow I'd say it's good for now an LGTM

----

As an aside, we should probably have a generic modal component that we can reuse instead of copying over the CSS from the profile modal into this modal the way we're doing. But we can make a new ticket for that

jlanesmith commented:
Okay sounds good! I don't have the authorization to merge it in, so I'll leave it for you!

scholvat commented:
@AngelaKrone could you get this released? We'll let you merge in to master, since master should always reflect production.

Good work Jonathan!

scholvat commented:
Merging into staging branch, to aggregate all the changes for the winter 2020 release
